### PR TITLE
Visualize Motor Fault Data + Thunderloop Safety Updates

### LINKED ISSUE #2895: Frequently check for motor faults in MotorService
### Description of the task
We should check for motor faults frequently and take appropriate action when a fault is noticed. E.g. When a motor has `shortdet_u` fault (may happen due to fast acceleration), it stops running and will not run again until the motor is resetted manually in code or thunderloop is restarted. It is important for us to automate this process (in a safe manner) s.t. during an actual game the robot can continue functioning without manual resetting. When other faults are noticed, we may want to shutdown the robot to avoid further damage.
Note that checking for faults requires SPI transfers which is slow on the Jetson, so to have thunderloop running at max frequency, we would want to spread out fault checks over different iterations of thunderloop, or possibly grouping SPI transfers. 

Observed faults should be added to the MotorStatus message. It would be nice to have some sort of red outline/flashing/... in Robot View widget to show when there is an `ErrorCode` or when there's a fault. This is really important functionality for all software and elec to have while field testing and debugging.

### Acceptance criteria

- [ ] Detect if any of the 5 motors fault without significantly slowing down thunderloop frequency
- [ ] Take appropriate action when a fault is detected
- [ ] Visualize fault/error code in Robot View widget

### PR DESCRIPTION
**Pull request by sauravbanna**
### Description

The newly added motor fault data to RobotStatus was not being visualized. This PR adds a new widget to Robot Status which displays this information

--------

This PR has various safety updates for Thunderloop:
- custom signal handling (addresses the issue where if a motor is running and thunderloop crashes, it still keeps running)
- encoder connection checks (address the problem with us burning motors by forgetting to plug in the encoder, via unsafe ramp up)
- Thunderloop changes to address #2895 

### Testing Done

Thunderscope builds and runs normally. Tests with manual data to see the different states the widget can be in
- signal handling (tested on a robot with various SIGINTs and SIGABRTs)
- tested encoder connection check by unplugging an encoder and ensuring that thunderloop crashes

### Resolved Issues

resolves #2895 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

**Comment by sauravbanna** [comment]
added some images but still tryna figure out how to add them to the tooltip, but rest is working

File: src/software/thunderscope/robot_diagnostics/robot_info.py [hunk 1]
```diff
@@ -69,24 +72,39 @@ def __init__(
...
+        self.breakbeam_label.setStyleSheet(
+            f"background-color: transparent; {self.BREAKBEAM_BORDER}"
+        )
 
+        self.robot_model_layout.addWidget(self.breakbeam_label)
+        self.robot_model_layout.addWidget(self.robot_model)
+        self.layout.addLayout(self.robot_model_layout)
+
+        self.layout.addLayout(self.status_layout)
+        # self.layout.setContentsMargins(0, 0, 0, 0)
```
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/robot_info.py]: nit: commented out code

**Review by nimazareian — COMMENTED** [review]
I just tried it and it looks good, though I still have to dig around the code a bit more to understand it. It could be helpful if we add a tooltip for the breakbeam as well. Just to make it more intuitive for other people not familiar with the UI. 

I was also curious if you're also planning to implement some sort of a feedback that shows that a robot has been disconnected? Somewhat similar to what we talked about before. (e.g. colors going grey, or a new connected widget changing color when the robot is connected/disconnected). And we can know if it's disconnected if we have not received a robot status message in the past ~1 second.

**Comment by sauravbanna** [comment]
> I just tried it and it looks good, though I still have to dig around the code a bit more to understand it. It could be helpful if we add a tooltip for the breakbeam as well. Just to make it more intuitive for other people not familiar with the UI.
> 
> I was also curious if you're also planning to implement some sort of a feedback that shows that a robot has been disconnected? Somewhat similar to what we talked about before. (e.g. colors going grey, or a new connected widget changing color when the robot is connected/disconnected). And we can know if it's disconnected if we have not received a robot status message in the past ~1 second.

yeah I can add a breakbeam tooltip, would it just display the current status (in / not in breakbeam) or any extra info?

and yea I was thinking of doing the robot disconnected stuff in a separate PR, I'll have that up soon

File: src/software/jetson_nano/services/motor.h [hunk 2]
```diff
@@ -169,39 +204,100 @@ class MotorService
...
 
     // SPI File Descriptors
-    std::unordered_map<int, int> file_descriptors;
+    std::unordered_map<int, int> file_descriptors_;
 
     // Drive Motors
-    EuclideanToWheel euclidean_to_four_wheel;
+    EuclideanToWheel euclidean_to_four_wheel_;
     std::unordered_map<int, bool> encoder_calibrated_;
+    std::unordered_map<int, MotorFaultIndicator> cached_motor_faults_;
```
Comment (itsarune) [inline src/software/jetson_nano/services/motor.h:281]: nima: nit: it seems like all the fields except a few don't have the trailing underscore. Lets make that consistent throughout

File: src/software/jetson_nano/services/motor.cpp [hunk 3]
```diff
@@ -898,3 +985,90 @@ void MotorService::startController(uint8_t motor, bool dribbler)
...
+        if (!calibrated_motors[motor])
+        {
+            calibrated = false;
+            LOG(WARNING) << "Encoder calibration check failure. " << MOTOR_NAMES[motor]
+                         << " did not change as expected";
+        }
+    }
+    if (!calibrated)
+    {
+        LOG(FATAL)
```
Comment (itsarune) [inline src/software/jetson_nano/services/motor.cpp:1081]: nima: nit: This right now prints only the name of the first motor that has its encoder calibrated, then crashes. It might be better to combine all names in the same LOG(FATAL), or have them as individual WARNINGs, and then a single FATAL message.

File: src/software/jetson_nano/services/motor.cpp [hunk 4]
```diff
@@ -210,106 +208,188 @@ bool MotorService::checkDriverFault(uint8_t motor)
...
-        LOG(DEBUG) << "RESET DETECTED";
-        setUpMotors();
-        encoders_calibrated = false;
+        is_initialized_ = false;
     }
-    // check if encoders are calibrated
-    if (!encoders_calibrated)
+
+    // checks if any motor has reset, sends a log message if so
+    for (uint8_t motor = 0; motor < NUM_MOTORS; ++motor)
```
Comment (itsarune) [inline src/software/jetson_nano/services/motor.cpp:396]: nima: Does encoder calibration and checking encoder_calibrated_ have to be done in poll? Can't it be done in setup? Since if I understand the logic correctly, we only calibrate at initialization and if a reset was detected.

File: src/software/jetson_nano/services/motor.cpp [hunk 5]
```diff
@@ -195,13 +191,15 @@ bool MotorService::checkDriverFault(uint8_t motor)
         LOG(WARNING)
             << "Indicates that the IC has been reset. All registers have been cleared to reset values."
             << "Attention: DRV_EN must be high to allow clearing reset";
+        motor_faults.insert(TbotsProto::MotorFault::RESET);
```
Comment (itsarune) [inline src/software/jetson_nano/services/motor.cpp:212]: nima: The one thing I've previously noticed is that the RESET bit always seems to be set to 1. So if we're frequently checking for faults, it will keep spamming the "==== FAULTS FOR MOTOR ..." log and this log. I'm not really familiar with why this bit is always set and if that's expected. If it is expected, then maybe we should remove their logs to reduce the spam

File: src/software/jetson_nano/services/motor.cpp [hunk 6]
```diff
@@ -210,106 +208,188 @@ bool MotorService::checkDriverFault(uint8_t motor)
...
+            TbotsProto::DriveUnit drive_status;
+            drive_status.set_enabled(cached_motor_faults_[motor].drive_enabled);
+
+            for (const TbotsProto::MotorFault& fault :
+                 cached_motor_faults_[motor].motor_faults)
+            {
+                drive_status.add_motor_faults(fault);
+            }
+
+            if (motor == FRONT_LEFT_MOTOR_CHIP_SELECT)
```
Comment (itsarune) [inline src/software/jetson_nano/services/motor.cpp:335]: nima: nit: Since we have the up to date wheel velocities for all wheels, I think this line should be moved out of the if statement so the proto always has the up to date velocity for all wheels.

File: src/software/jetson_nano/services/motor.cpp [hunk 7]
```diff
@@ -210,106 +208,188 @@ bool MotorService::checkDriverFault(uint8_t motor)
...
+{
+    TbotsProto::MotorStatus motor_status;
+
+    cached_motor_faults_[motor_fault_detector_] = checkDriverFault(motor_fault_detector_);
+
+    for (uint8_t motor = 0; motor < NUM_MOTORS; ++motor)
+    {
+        if (motor != DRIBBLER_MOTOR_CHIP_SELECT)
+        {
+            TbotsProto::DriveUnit drive_status;
```
Comment (itsarune) [inline src/software/jetson_nano/services/motor.cpp:326]: nima: This for loop seems unnecessary, since we're not using i and we're updating 1 wheel with each iteration as well

File: src/software/jetson_nano/services/motor.cpp [hunk 8]
```diff
@@ -177,11 +167,17 @@ void MotorService::setUpMotors()
...
 
+    checkEncoderConnections();
 
-bool MotorService::checkDriverFault(uint8_t motor)
+    is_initialized_ = true;
+}
+
+MotorService::MotorFaultIndicator MotorService::checkDriverFault(uint8_t motor)
 {
+    bool drive_enabled = true;
```
Comment (itsarune) [inline src/software/jetson_nano/services/motor.cpp:196]: nima: If drive is not enabled for some motor, do we want to reset it? Or is the plan to manually reset it by restarting thunderloop?

File: src/software/jetson_nano/services/motor.h [hunk 9]
```diff
@@ -82,12 +83,46 @@ class MotorService
...
+
    private:
     /**
-     * Checks for faults, clears them and sets up motors.
-     *
+     * Holds motor fault information for a particular motor and whether any fault has
+     * caused the motor to be disabled.
      */
-    void setUpMotors();
+    struct MotorFaultIndicator
```
Comment (itsarune) [inline src/software/jetson_nano/services/motor.h:102]: nima: Is there a reason why we can't use the existing TbotsProto::DriveUnit instead of this?

File: src/software/jetson_nano/thunderloop.cpp [hunk 10]
```diff
@@ -19,6 +19,31 @@
...
+     *
+     * @param the signal value (SIGINT, SIGABRT, SIGTERN, etc)
+     */
+    void tbotsExit(int signal_num)
+    {
+        g_motor_service->resetMotorBoard();
+
+        // by now g3log may have died due to the termination signal, so it isn't reliable
+        // to log messages
+        std::cerr << "\n\n!!!\nReceived termination signal: "
```
Comment (itsarune) [inline src/software/jetson_nano/thunderloop.cpp:38]: nima: From my understanding, \n doesn't guarantee that the cout buffer is flushed, so it might be safer to use std::endl to make sure it is flushed. A better alternative could be to use std::cerr instead, which logs to the error stream which is not buffered.

**Comment by itsarune** [comment]
for posterity, I copied over @nimazareian's comments

File: src/software/jetson_nano/services/motor.h (hunk 2 above)
Comment (itsarune) [reply src/software/jetson_nano/services/motor.h:281]: made all class members with trailing underscore @nimazareian

File: src/software/jetson_nano/services/motor.cpp (hunk 3 above)
Comment (itsarune) [reply src/software/jetson_nano/services/motor.cpp:1081]: each motor is logged with a warning now and then a final fatal message @nimazareian

File: src/software/jetson_nano/services/motor.cpp (hunk 4 above)
Comment (itsarune) [reply src/software/jetson_nano/services/motor.cpp:396]: I moved the calibration to setup() @nimazareian

File: src/software/jetson_nano/services/motor.cpp (hunk 5 above)
Comment (itsarune) [reply src/software/jetson_nano/services/motor.cpp:212]: I added code to clear the RESET flag to setup() @nimazareian

File: src/software/jetson_nano/services/motor.cpp (hunk 6 above)
Comment (itsarune) [reply src/software/jetson_nano/services/motor.cpp:335]: I moved it

File: src/software/jetson_nano/services/motor.cpp (hunk 7 above)
Comment (itsarune) [reply src/software/jetson_nano/services/motor.cpp:326]: there was a mistake initially in this code, the iterator variable is used now

File: src/software/jetson_nano/services/motor.cpp (hunk 8 above)
Comment (itsarune) [reply src/software/jetson_nano/services/motor.cpp:196]: yeah my code doesn't handle this other than just notifying thunderscope

Comment (itsarune) [reply src/software/jetson_nano/services/motor.cpp:196]: I think thunderloop should be responsible for resetting on faults, what do you think? @nimazareian

File: src/software/jetson_nano/services/motor.h (hunk 9 above)
Comment (itsarune) [reply src/software/jetson_nano/services/motor.h:102]: I thought about this at first, but I thought it would be awkward to have partially-filled protos in checkDriverFault, which is why I made interim structs

File: src/software/jetson_nano/thunderloop.cpp (hunk 10 above)
Comment (itsarune) [reply src/software/jetson_nano/thunderloop.cpp:38]: changed to std::cerr @nimazareian

File: src/software/jetson_nano/services/motor.h (hunk 9 above)
Comment (itsarune) [reply src/software/jetson_nano/services/motor.h:102]: @nimazareian

File: src/software/jetson_nano/services/motor.cpp (hunk 8 above)
Comment (itsarune) [reply src/software/jetson_nano/services/motor.cpp:196]: @nimazareian

File: src/software/jetson_nano/services/motor.cpp (hunk 7 above)
Comment (itsarune) [reply src/software/jetson_nano/services/motor.cpp:326]: @nimazareian

File: src/software/jetson_nano/services/motor.cpp (hunk 6 above)
Comment (itsarune) [reply src/software/jetson_nano/services/motor.cpp:335]: @nimazareian

File: src/software/thunderscope/common/common_widgets.py [hunk 11]
```diff
@@ -382,3 +382,42 @@ def draw_robot(painter, rect, start_angle_degree, span_angle_degree):
     painter.drawChord(
         rect, start_angle_degree * convert_degree, span_angle_degree * convert_degree,
     )
+
+
+def get_string_val(descriptor, value):
```
Comment (itsarune) [inline src/software/thunderscope/common/common_widgets.py]: why did you move this to common_widget.py from proto_parameter_util?

File: src/software/thunderscope/robot_diagnostics/robot_info.py [hunk 12]
```diff
@@ -6,6 +6,61 @@
...
+        self.update_breakbeam_status(None)
+
+    def update_breakbeam_status(self, new_breakbeam_status):
+        """
+        Updates the current breakbeam status and refreshes the UI accordingly
+        :param new_breakbeam_status: the new breakbeam status
+        """
+        self.breakbeam_status = new_breakbeam_status
+
+        if self.breakbeam_status == None:
```
Comment (itsarune) [inline src/software/thunderscope/robot_diagnostics/robot_info.py]: nit:
```suggestion
        if self.breakbeam_status is None:
```

File: src/software/thunderscope/robot_diagnostics/robot_info.py [hunk 13]
```diff
@@ -186,24 +251,26 @@ def create_vision_pattern_label(self, team_colour, radius):
...
-            self.breakbeam_label.setStyleSheet("background-color: red")
-        else:
-            self.breakbeam_label.setText("Not in Beam")
-            self.breakbeam_label.setStyleSheet("background-color: green")
+        self.breakbeam_label.update_breakbeam_status(power_status.breakbeam_tripped)
+
+        self.motor_fault_view.refresh(
+            motor_status,
+            motor_status.front_left.DESCRIPTOR.fields_by_name["motor_faults"],
+        )
```
Comment (itsarune) [inline src/software/thunderscope/robot_diagnostics/robot_info.py]: looks like you're only checking front_left

File: src/software/thunderscope/robot_diagnostics/robot_view.py [hunk 14]
```diff
@@ -123,7 +127,27 @@ def refresh(self):
...
+                back_right=DriveUnit(
+                    motor_faults=[MotorFault.DRIVER_OVERTEMPERATURE_PREWARNING],
+                    enabled=False,
+                ),
+                dribbler=DribblerStatus(
+                    motor_faults=[MotorFault.DRIVER_OVERTEMPERATURE_PREWARNING],
+                    enabled=True,
+                ),
+            ),
+        )
```
Comment (itsarune) [inline src/software/thunderscope/robot_diagnostics/robot_view.py]: this looks like it's meant for testing

**Review by itsarune — COMMENTED** [review]
looking good ->

File: src/software/jetson_nano/services/motor.cpp (hunk 8 above)
Comment (nimazareian) [reply src/software/jetson_nano/services/motor.cpp:196]: I think we should reset the motor, but it might also be a good idea to discuss the implications of resetting a faulty motor driver instantly with Elec, and if there are any risks involved.

File: src/software/thunderscope/robot_diagnostics/BUILD [hunk 15]
```diff
@@ -69,3 +79,14 @@ py_library(
...
+py_library(
+    name = "motor_fault_icon_loader",
+    srcs = ["motor_fault_icons/motor_fault_icon_loader.py"],
+    data = [
+        "motor_fault_icons/no_fault.png",
+        "motor_fault_icons/stopped.png",
+        "motor_fault_icons/warning.png",
+    ],
+    deps = [],
+)
```
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/BUILD]: Nit: For consistency with the rest of our code base, all folders should have their own BUILD file. So this should be moved to the `motor_faults_icons` folder

File: src/software/thunderscope/robot_diagnostics/motor_fault_view.py [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+                )
+                self.motor_fault_display.setText("Warning" if self.enabled else "Error")
+
+    def refresh(self, message, enum_descriptor):
+        """
+        Converts the given message into a map of motor name to its fault info
+        And calls functions to update the main UI and the tooltip
+        :param message: the message to update the widget with
+        :param enum_descriptor: descriptor to translate from enum indexes to string values
+        """
```
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/motor_fault_view.py]: nit: is `message` supposed to be a `motor_status` message? If so, could you clarify that by either adding the variable type, or updating the variable's name?

File: src/software/thunderscope/robot_diagnostics/motor_fault_view.py [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+            self.motor_fault_display.setStyleSheet("background: green;")
+            self.motor_fault_display.setText("No Fault")
+
+            if self.fault_count:
+                self.motor_fault_display.setStyleSheet(
+                    "background: yellow; color: black;"
+                    if self.enabled
+                    else "background: red;"
+                )
+                self.motor_fault_display.setText("Warning" if self.enabled else "Error")
```
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/motor_fault_view.py]: nit: the logic here can be slightly simplified into a single if else statement
```suggestion
            if self.fault_count > 0:
                self.fault_count_label.setText(str(self.fault_count))
                self.fault_count_label.show()
                
                self.motor_fault_display.setStyleSheet(
                    "background: yellow; color: black;"
                    if self.enabled
                    else "background: red;"
                )
                self.motor_fault_display.setText("Warning" if self.enabled else "Error")
            else:
                self.fault_count_label.hide()

                self.motor_fault_display.setStyleSheet("background: green;")
                self.motor_fault_display.setText("No Fault")
```

File: src/software/thunderscope/robot_diagnostics/robot_info.py [hunk 18]
```diff
@@ -6,6 +6,61 @@
...
+    def event(self, event):
+        """
+        Overridden event function which intercepts all events
+        On hover, displays a tooltip with the current breakbeam status
+        :param event: event to check
+        """
+        common_widgets.display_tooltip(
+            event,
+            "No Signal Yet"
+            if self.breakbeam_status == None
```
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/robot_info.py]: To stay consistent with Arun's comment
```suggestion
            if self.breakbeam_status is None
```

File: src/software/thunderscope/robot_diagnostics/robot_info.py (hunk 13 above)
Comment (nimazareian) [reply src/software/thunderscope/robot_diagnostics/robot_info.py]: I digged around a bit more and it looks like this line is just to get the DESCRIPTOR for the repeated motor fault enum type. It would be a good idea to add a comment clarifying that the line is to get the descriptor rather than the actual field value.

**Review by nimazareian — COMMENTED** [review]
Overall looks good. Tried to review all of the parts that seemed to be new since my last review.

File: src/software/thunderscope/common/common_widgets.py (hunk 11 above)
Comment (sauravbanna) [reply src/software/thunderscope/common/common_widgets.py]: we need this to translate between
- the geneva enum values and other enum values in the parameter tree and the robot status tree
- the motor fault enum values in the motor fault widget

so since we need it in 2 places I thought I'd move it out

File: src/software/thunderscope/robot_diagnostics/robot_info.py (hunk 13 above)
Comment (sauravbanna) [reply src/software/thunderscope/robot_diagnostics/robot_info.py]: yeah this is just to get the descriptor, I'll add a comment explaining that

File: src/software/thunderscope/robot_diagnostics/robot_view.py (hunk 14 above)
Comment (sauravbanna) [reply src/software/thunderscope/robot_diagnostics/robot_view.py]: oops sorry about that

**Review by nimazareian — DISMISSED** [review]
LGTM!

File: src/software/jetson_nano/services/motor.cpp (hunk 8 above)
Comment (sauravbanna) [reply src/software/jetson_nano/services/motor.cpp:196]: arun: we'll handle this in a new PR

**Review by nimazareian — DISMISSED** [review]

File: src/software/thunderscope/common/common_widgets.py (hunk 11 above)
Comment (itsarune) [reply src/software/thunderscope/common/common_widgets.py]: but isn't it already in the thunderscope/common folder already? it seems like proto_parameter_util is already a shared set of helper functions

File: src/software/thunderscope/common/common_widgets.py [hunk 19]
```diff
@@ -382,3 +382,21 @@ def draw_robot(painter, rect, start_angle_degree, span_angle_degree):
...
+    Upon Enter, displays a tooltip with the given text
+    Upon Leave, hides the tooltip
+    :param event: event to check
+    :param tooltip_text: the text to display in the tooltip
+    """
+    if str(event.type()) == "Type.Enter":
+        QToolTip.showText(
+            QPoint(int(event.globalPosition().x()), int(event.globalPosition().y()),),
+            tooltip_text,
+            msecShowTime=20000,
```
Comment (LiCody) [inline src/software/thunderscope/common/common_widgets.py]: can we change this to 20 * MILLISECOND_PER_SECOND

File: src/proto/robot_status_msg.proto [hunk 20]
```diff
@@ -72,10 +75,32 @@ enum ErrorCode
     DRIBBLER_MOTOR_HOT = 4;
 }
 
+enum MotorFault
+{
+    RESET                             = 0;
+    DRIVER_OVERTEMPERATURE_PREWARNING = 1;
```
Comment (LiCody) [inline src/proto/robot_status_msg.proto:84]: can we link to the trinamic doc with what each is

**Review by LiCody — COMMENTED** [review]
2 nits, LGTM

**Review by LiCody — COMMENTED** [review]
2 nits, LGTM

File: src/proto/robot_status_msg.proto (hunk 20 above)
Comment (sauravbanna) [reply src/proto/robot_status_msg.proto:84]: @itsarune you would probably know more about this? i'll try to find docs for this in the meantime

File: src/software/thunderscope/common/common_widgets.py (hunk 19 above)
Comment (sauravbanna) [reply src/software/thunderscope/common/common_widgets.py]: done

File: src/proto/robot_status_msg.proto (hunk 20 above)
Comment (itsarune) [reply src/proto/robot_status_msg.proto:84]: updated

**Review by nimazareian — APPROVED** [review]

**Review by LiCody — APPROVED** [review]
